pkg/mutator/ratelimit.go

```diff
@@ -0,0 +1,99 @@
+package mutator
+
+import (
+	"fmt"
+	"math"
+
+	sensor "github.com/argoproj/argo-events/pkg/apis/sensor/v1alpha1"
+)
+
+const (
+	secondsInHour = 3600
+	minutesInHour = 60
+)
+
+type RateLimitCalculator struct {
+	defaultRateLimit sensor.RateLimit
+}
+
+func NewRateLimitCalculator(defaultRequestsPerHour int32) *RateLimitCalculator {
+	return &RateLimitCalculator{
+		defaultRateLimit: sensor.RateLimit{
+			Unit:            sensor.Hour,
+			RequestsPerUnit: defaultRequestsPerHour,
+		},
+	}
+}
+
+// Calculates the RateLimit based on min(sensorValue, maxRateLimit) where
+// maxRateLimit is the namespaceValue if set, otherwise defaultRateLimit.
+// Returns a value in requests per hour.
+func (r *RateLimitCalculator) Calculate(namespaceValue, sensorValue *sensor.RateLimit) (sensor.RateLimit, error) {
+	maxRateLimit := r.defaultRateLimit
+	if namespaceValue != nil {
+		// Namespace value overrides the default
+		maxRateLimit = *namespaceValue
+	}
+
+	normalizedMaxRateLimit, err := normalizeToHour(maxRateLimit)
+	if err != nil {
+		return sensor.RateLimit{}, err
+	}
+
+	if sensorValue == nil {
+		return normalizedMaxRateLimit, nil
+	}
+
+	// Compare normalized values
+	normalizedSensorValue, err := normalizeToHour(*sensorValue)
+	if err != nil {
+		return sensor.RateLimit{}, err
+	}
+
+	return min(normalizedMaxRateLimit, normalizedSensorValue)
+}
+
+func normalizeToHour(input sensor.RateLimit) (sensor.RateLimit, error) {
```

Comment:
We need to be able to clearly communicate defaults and what behavior to expect for the end user. 

What do we want to happen for the following rate limits:

default: unit: second, rate: 10

User requests:
1. unit: Hour, rate: 3600
2. unit: Minute, rate: 60
3. unit: Second, rate: 1

These requests normalize to the same req/hour rate but do 1 and 2 pass the default rate limit?

Reply:
Currently the code normalizes all values to Hours.

So using your example, default becomes unit: hour, rate: 36000

User requests:
1. unit: Hour, rate: 3600 -> passes default
2. unit: Minute, rate: 60 -> gets normalized to hour, same result
3. unit: Second, rate: 1 -> gets normalized to hour, same result

One thing that could be changed is that if the user request is acceptable, we keep it in the user's unit, so there's no confusion of why the unit got modified.

Reply:
I don't know that we want to accept 1 or 2 though because they can exceed our desired per second rate. I suppose we don't know what the actual rate values we want but making 3600 pods in a few seconds is not desired and is allowed for an hourly rate limit. 

Thoughts?

Reply:
Mmm that's a really good point, I completely missed that. Let me digest that a bit and get back to you.

Reply:
What if we base the rate limit down to the lowest unit we want to tolerate?

Reply:
(already discussed offline, but capturing here for visibility)

The [uber-go/ratelimit](https://github.com/uber-go/ratelimit) lib that argo-events [uses](https://github.com/argoproj/argo-events/blob/master/sensors/listener.go#L97) is extremely simple. RateLimit of 1 per second, 60 per minute, 3600 per hour are equivalent and all wait for 1s per request.
See this [playground](https://go.dev/play/p/3lSlpGAJ3os) demonstration. Uncomment lines 13 to 15 selectively to see you can’t Take() in bursts, it’s the normalized per second interval.
I haven’t tried with an actual argo-events Sensor resource, but from the [code](https://github.com/argoproj/argo-events/blob/master/sensors/listener.go#L356) it’s the same thing.

So we don't need to worry about the units meaning different burst rates.

Some changes in the latest commit:
- allow both default `unit` and `requests-per-unit` to be set via command line
- the calculation of the minimum is now done using Second as the base unit so no more worries about overflow
- the value returned from `Calculate` is the unaltered [sensorValue, namespaceValue, or defaultValue]. Previously I was normalizing it. Unaltered is better so it's clear to users their value persists if it's below the limit, and over the limit our defaults trump theirs.